File: problems/0008_string-to-integer-atoi/built-in/Solution.scala
object Solution {
    def myAtoi(str: String): Int = {
        str match {
            case NUM_FMT(intStr) =>
                val result = BigInt(intStr)
                if (result < Int.MinValue)
                    Int.MinValue
                else if (result > Int.MaxValue)
                    Int.MaxValue
                else
                    result.toInt
            case _ =>
                0
        }
    }
    val NUM_FMT = """\s*([\-\+]?\d+).*""".r
}

File: problems/0008_string-to-integer-atoi/built-in/Solution.kt
class Solution {
    fun myAtoi(str: String): Int {
        val matches = NUM_FMT.find(str)
        if (matches != null)
            try {
                val (s) = matches!!.destructured
                val result = s.toBigInteger()
                if (result < INT_MIN)
                    return Int.MIN_VALUE
                else if (result > INT_MAX)
                    return Int.MAX_VALUE
                else
                    return result.toInt()
            } catch (ignored: NumberFormatException) {
            }
        return 0
    }
    val INT_MIN = Int.MIN_VALUE.toBigInteger()
    val INT_MAX = Int.MAX_VALUE.toBigInteger()
    val NUM_FMT = """^\s*([\-\+]?\d+).*$""".toRegex()
}
